File: use.py
from globals import *

__all__ = ['quaff', 'read_scroll', 'vanish', 'identify', 'eat', 'teleport',
           'hallucinate', 'unhallucinate', 'unblind', 'confuse', 'unconfuse']

def quaff():
    ch = get_pack_letter("quaff what? ", POTION)
    if ch == CANCEL:
        return
    obj = get_letter_object(ch)
    if not obj:
        message("no such item.", 0)
        return
    if obj.what_is != POTION:
        message("you can't drink that", 0)
        return
    k = obj.which_kind
    if k == INCREASE_STRENGTH:
        message("you feel stronger now, what bulging muscles!", 0)
        rogue.strength_current += 1
        if rogue.strength_current > rogue.strength_max:
            rogue.strength_max = rogue.strength_current
    elif k == RESTORE_STRENGTH:
        message("this tastes great, you feel warm all over", 0)
        rogue.strength_current = rogue.strength_max
    elif k == HEALING:
        message("you begin to feel better", 0)
        potion_heal(0)
    elif k == EXTRA_HEALING:
        message("you begin to feel much better", 0)
        potion_heal(1)
    elif k == POISON:
        rogue.strength_current -= get_rand(1, 3)
        if rogue.strength_current < 0:
            rogue.strength_current = 0
        message("you feel very sick now", 0)
        if g.halluc:
            unhallucinate()
    elif k == RAISE_LEVEL:
        message("you feel more experienced", 0)
        add_exp(level_points[rogue.exp - 1] - rogue.exp_points + 1)
    elif k == BLINDNESS:
        go_blind()
    elif k == HALLUCINATION:
        message("oh wow, everything seems so cosmic", 0)
        g.halluc += get_rand(500, 800)
    elif k == DETECT_MONSTER:
        if g.level_monsters.next_object:
            show_monsters()
        else:
            message("you have a strange feeling for a moment, then it passes", 0)
        g.detect_monster = 1
    elif k == DETECT_OBJECTS:
        if g.level_objects.next_object:
            if not g.blind:
                show_objects()
        else:
            message("you have a strange feeling for a moment, then it passes", 0)
    elif k == CONFUSION:
        message("what a trippy feeling" if g.halluc else "you feel confused", 0)
        confuse()
    print_stats()
    if id_potions[k].id_status != CALLED:
        id_potions[k].id_status = IDENTIFIED
    vanish(obj, 1)

def read_scroll():
    ch = get_pack_letter("read what? ", SCROLL)
    if ch == CANCEL:
        return
    obj = get_letter_object(ch)
    if not obj:
        message("no such item.", 0)
        return
    if obj.what_is != SCROLL:
        message("you can't read that", 0)
        return
    k = obj.which_kind
    if k == SCARE_MONSTER:
        message("you hear a maniacal laughter in the distance", 0)
    elif k == HOLD_MONSTER:
        hold_monster()
    elif k == ENCHANT_WEAPON:
        if rogue.weapon:
            message("your %sglows %sfor a moment" % (id_weapons[rogue.weapon.which_kind].title, get_ench_color()), 0)
            if get_rand(0, 1):
                rogue.weapon.to_hit_enchantment += 1
            else:
                rogue.weapon.damage_enchantment += 1
            rogue.weapon.is_cursed = 0
        else:
            message("your hands tingle", 0)
    elif k == ENCHANT_ARMOR:
        if rogue.armor:
            message("your armor glows %sfor a moment" % get_ench_color(), 0)
            rogue.armor.damage_enchantment += 1
            rogue.armor.is_cursed = 0
            print_stats()
        else:
            message("your skin crawls", 0)
    elif k == IDENTIFY:
        message("this is a scroll of identify", 0)
        message("what would you like to identify?", 0)
        obj.identified = 1
        id_scrolls[k].id_status = IDENTIFIED
        identify()
    elif k == TELEPORT:
        teleport()
    elif k == SLEEP:
        sleep_scroll()
    elif k == PROTECT_ARMOR:
        if rogue.armor:
            message( "your armor is covered by a shimmering gold shield", 0)
            rogue.armor.is_protected = 1
        else:
            message("your acne seems to have disappeared", 0)
    elif k == REMOVE_CURSE:
        message("you feel as though someone is watching over you", 0)
        if rogue.armor:
            rogue.armor.is_cursed = 0
        if rogue.weapon:
            rogue.weapon.is_cursed = 0
    elif k == CREATE_MONSTER:
        create_monster()
    elif k == AGGRAVATE_MONSTER:
        aggravate()
    if id_scrolls[k].id_status != CALLED:
        id_scrolls[k].id_status = IDENTIFIED
    vanish(obj, 1)
    
def vanish(obj, rm):
    if obj.quantity > 1:
        obj.quantity -= 1
    else:
        remove_from_pack(obj, rogue.pack)
        make_avail_ichar(obj.ichar)
    if rm:
        register_move()
    
def potion_heal(extra):
    ratio = float(rogue.hp_current) / rogue.hp_max
    if ratio >= 0.9:
        rogue.hp_max += extra + 1
        rogue.hp_current = rogue.hp_max
    else:
        if ratio < 30.0:
            ratio = 30.0
        if extra:
            ratio += ratio
        add = int(ratio * (rogue.hp_current - rogue.hp_max))
        rogue.hp_current = max(rogue.hp_current + add, rogue.hp_max)
    if g.blind:
        unblind()
    if g.confused and extra:
        unconfuse()
    elif g.confused:
        g.confused = (g.confused - 9)  / 2
        if g.confused <= 0:
            unconfuse()
    if g.halluc and extra:
        unhallucinate()
    elif g.halluc:
        g.halluc = g.halluc / 2 + 1
  
def identify():
    while True:
        ch = get_pack_letter("identify what? ", IS_OBJECT)
        if ch == CANCEL:
            return
        obj = get_letter_object(ch)
        if not obj:
            message("no such item, try again", 0)
            check_message()
            continue
        obj.identified = 1
        if obj.what_is & (SCROLL | POTION | WEAPON | ARMOR | WAND):
            id_table = get_id_table(obj)
            id_table[obj.which_kind].id_status = IDENTIFIED
        message(get_description(obj), 0)
        return

def eat():
    ch = get_pack_letter("eat what? ", FOOD)
    if ch == CANCEL:
        return
    obj = get_letter_object(ch)
    if not obj:
        message("no such item.", 0)
        return
    if obj.what_is != FOOD:
        message("you can't eat that", 0)
        return
    moves = get_rand(800, 1000)
    if moves >= 900:
        message("yum, that tasted good", 0)
    else:
        message("yuk, that food tasted awful", 0)
        add_exp(3)
    rogue.moves_left /= 2
    rogue.moves_left += moves
    g.hunger_str = ""
    print_stats()
    
    vanish(obj, 1)

def hold_monster():
    mcount = 0
    for i in range(-2, 3):
        for j in range(-2, 3):
            row = rogue.row + i
            col = rogue.col + j
            if row < MIN_ROW or row > LINES - 2 or col < 0 or col > COLS - 1:
                continue
            if screen[row][col] & MONSTER:
                monster = object_at(g.level_monsters, row, col)
                monster.m_flags |= IS_ASLEEP
                monster.m_flags &= ~WAKENS
                mcount += 1
    if mcount == 0:
        message("you feel a strange sense of loss", 0)
    elif mcount == 1:
        message("the monster freezes", 0)
    else:
        message("the monsters around you freeze", 0)

def teleport():
    if g.current_room >= 0:
        darken_room(g.current_room)
    else:
        mvaddch(rogue.row, rogue.col, get_room_char(screen[rogue.row][rogue.col], rogue.row, rogue.col))
    put_player()
    light_up_room()
    g.being_hold = 0
    
def hallucinate():
    if g.blind:
        return
    obj = g.level_objects.next_object
    while obj:
        ch = mvinch(obj.row, obj.col)
        if (ch < 'A' or ch > 'Z') and (obj.row != rogue.row or obj.col != rogue.col):
            if ch != ' ' and ch != '.' and ch != '#' and ch != '+':
                addch(get_rand_obj_char())
        obj = obj.next_object
    
    obj = g.level_monsters.next_object
    while obj:
        ch = mvinch(obj.row, obj.col)
        if ch >= 'A' and ch <= 'Z':
            addch(chr(get_rand(ord('A'), ord('Z'))))
        obj = obj.next_object

def unhallucinate():
    g.halluc = 0
    if g.current_room == PASSAGE:
        light_passage(rogue.row, rogue.col)
    else:
        light_up_room()
    message("everything looks SO boring now", 0)

def unblind():
    g.blind = 0
    message("the veil of darkness lifts", 0)
    if g.current_room == PASSAGE:
        light_passage(rogue.row, rogue.col)
    else:
        light_up_room()
    if g.detect_monster:
        show_monsters()
    if g.halluc:
        hallucinate()
    
def sleep_scroll():
    message("you fall asleep", 0)
    ##sleep(1)
    i = get_rand(4, 10)
    while i:
        move_monsters()
        i -= 1
    ##sleep(1)
    message("you can move again", 0)

def go_blind():
    if not g.blind:
        message("a cloak of darkness falls around you", 0)
    g.blind += get_rand(500, 800)
    
    if g.current_room >= 0:
        r = rooms[g.current_room]
        for i in range(r.top_row + 1, r.bottom_row):
            for j in range(r.left_col + 1, r.right_col):
                mvaddch(i, j, ' ')
    mvaddch(rogue.row, rogue.col, rogue.fchar)
    refresh()

def get_ench_color():
    if g.halluc:
        return id_potions[get_rand(0, POTIONS - 1)].title
    return "blue "

def confuse():
    g.confused = get_rand(12, 22)

def unconfuse():
    g.confused = 0
    message("you feel less %s now" % ("trippy" if g.halluc else "confused"), 0)

File: flrogue/lib/rogue/use.dart
import 'globals.dart';
import 'level.dart' hide levelPoints;
import 'message.dart';
import 'monster.dart';
import 'move.dart';
import 'object.dart';
import 'pack.dart';
import 'room.dart';
import 'ui.dart';

Future<void> quaff() async {
  String ch = await getPackLetter("quaff what?", Cell.potion);
  if (ch == cancel) {
    return;
  }

  GameObject? obj = getLetterObject(ch);
  if (obj == null) {
    await message("no such item.");
    return;
  }

  if (obj.whatIs != Cell.potion) {
    await message("you can't drink that");
    return;
  }

  int k = obj.whichKind;
  if (k == PotionType.increaseStrength.index) {
    await message("you feel stronger now, what bulging muscles!");
    rogue.strengthCurrent += 1;
    if (rogue.strengthCurrent > rogue.strengthMax) {
      rogue.strengthMax = rogue.strengthCurrent;
    }
  } else if (k == PotionType.restoreStrength.index) {
    await message("this tastes great, you feel warm all over");
    rogue.strengthCurrent = rogue.strengthMax;
  } else if (k == PotionType.healing.index) {
    await message("you begin to feel better");
    await potionHeal(false);
  } else if (k == PotionType.extraHealing.index) {
    await message("you begin to feel much better");
    await potionHeal(true);
  } else if (k == PotionType.poison.index) {
    rogue.strengthCurrent -= getRand(1, 3);
    if (rogue.strengthCurrent < 0) {
      rogue.strengthCurrent = 0;
    }
    await message("you feel very sick now");
    if (halluc != 0) {
      await unhallucinate();
    }
  } else if (k == PotionType.raiseLevel.index) {
    await message("you feel more experienced");
    await addExp(levelPoints[rogue.exp - 1] - rogue.expPoints + 1);
  } else if (k == PotionType.blindness.index) {
    await goBlind();
  } else if (k == PotionType.hallucination.index) {
    await message("oh wow, everything seems so cosmic");
    halluc += getRand(500, 800);
  } else if (k == PotionType.detectMonster.index) {
    if (levelMonsters.isNotEmpty) {
      showMonsters();
    } else {
      await message("you have a strange feeling for a moment, then it passes");
    }
    detectMonster = true;
  } else if (k == PotionType.detectObjects.index) {
    if (levelObjects.isNotEmpty) {
      if (blind == 0) {
        showObjects();
      }
    } else {
      await message("you have a strange feeling for a moment, then it passes");
    }
  } else if (k == PotionType.confusion.index) {
    await message(halluc != 0 ? "what a trippy feeling" : "you feel confused");
    confuse();
  }

  printStats();

  if (idPotions[k].idStatus != IdStatus.called) {
    idPotions[k].idStatus = IdStatus.identified;
  }

  await vanish(obj, true);
}

Future<void> readScroll() async {
  String ch = await getPackLetter("read what?", Cell.scroll);
  if (ch == cancel) {
    return;
  }

  GameObject? obj = getLetterObject(ch);
  if (obj == null) {
    await message("no such item.");
    return;
  }

  if (obj.whatIs != Cell.scroll) {
    await message("you can't read that");
    return;
  }

  var k = ScrollType.values[obj.whichKind];
  switch (k) {
    case ScrollType.scareMonster:
      await message("you hear a maniacal laughter in the distance");
    case ScrollType.holdMonster:
      await holdMonster();
    case ScrollType.enchantWeapon:
      if (rogue.weapon case final weapon?) {
        await message(
          "your ${nameOf(weapon)}glow${weapon.quantity > 1 ? '' : 's'} ${getEnchColor()}for a moment",
        );
        if (getRand(0, 1) != 0) {
          weapon.toHitEnchantment += 1;
        } else {
          weapon.damageEnchantment += 1;
        }
        weapon.isCursed = 0;
      } else {
        await message("your hands tingle");
      }
    case ScrollType.enchantArmor:
      if (rogue.armor case final armor?) {
        await message("your armor glows ${getEnchColor()}for a moment");
        armor.damageEnchantment += 1;
        armor.isCursed = 0;
        printStats();
      } else {
        await message("your skin crawls");
      }
    case ScrollType.identify:
      await message("this is a scroll of identify");
      await message("what would you like to identify?");
      obj.identified = 1;
      idScrolls[k.index].idStatus = IdStatus.identified;
      await identify();
    case ScrollType.teleport:
      teleport();
    case ScrollType.sleep:
      await sleepScroll();
    case ScrollType.protectArmor:
      if (rogue.armor case final armor?) {
        await message("your armor is covered by a shimmering gold shield");
        armor.isProtected = 1;
        armor.isCursed = 0;
      } else {
        await message("your acne seems to have disappeared");
      }
    case ScrollType.removeCurse:
      await message(
        halluc == 0
            ? "you feel as though someone is watching over you"
            : "you feel in touch with the universal oneness",
      );
      rogue.armor?.isCursed = 0;
      rogue.weapon?.isCursed = 0;
    case ScrollType.createMonster:
      await createMonster();
    case ScrollType.aggravateMonster:
      await aggravate();
    case ScrollType.magicMapping:
      await message("this scroll seems to have a map on it");
      drawMagicMap();
  }

  if (idScrolls[k.index].idStatus != IdStatus.called) {
    idScrolls[k.index].idStatus = IdStatus.identified;
  }

  await vanish(obj, true);
}

Future<void> vanish(GameObject obj, bool rm) async {
  if (obj.quantity > 1) {
    obj.quantity -= 1;
  } else {
    removeFromPack(obj, rogue.pack);
    makeAvailIchar(obj.ichar);
  }

  if (rm) {
    await registerMove();
  }
}

Future<void> potionHeal(bool extra) async {
  double ratio = rogue.hpCurrent / rogue.hpMax;

  if (ratio >= 0.9) {
    rogue.hpMax += extra ? 2 : 1;
    rogue.hpCurrent = rogue.hpMax;
  } else {
    if (ratio < 0.33) {
      ratio = 0.33;
    }

    if (extra) {
      ratio += ratio;
    }

    int add = (ratio * (rogue.hpMax - rogue.hpCurrent)).toInt();
    rogue.hpCurrent = [
      rogue.hpCurrent + add,
      rogue.hpMax,
    ].reduce((a, b) => a > b ? a : b);
  }

  if (blind != 0) {
    await unblind();
  }

  if (confused != 0 && extra) {
    await unconfuse();
  } else if (confused != 0) {
    confused = (confused - 9) ~/ 2;
    if (confused <= 0) {
      await unconfuse();
    }
  }

  if (halluc != 0 && extra) {
    await unhallucinate();
  } else if (halluc != 0) {
    halluc = halluc ~/ 2 + 1;
  }
}

Future<void> identify() async {
  while (true) {
    String ch = await getPackLetter("identify what?", Cell.isObject);
    if (ch == cancel) {
      return;
    }

    GameObject? obj = getLetterObject(ch);
    if (obj == null) {
      await message("no such item, try again");
      checkMessage();
      continue;
    }

    obj.identified = 1;
    if (obj.whatIs &
            (Cell.scroll |
                Cell.potion |
                Cell.weapon |
                Cell.armor |
                Cell.wand) !=
        0) {
      List<Identity> idTable = getIdTable(obj);
      idTable[obj.whichKind].idStatus = IdStatus.identified;
    }

    await message(getDescription(obj));
    return;
  }
}

Future<void> eat() async {
  String ch = await getPackLetter("eat what?", Cell.food);
  if (ch == cancel) {
    return;
  }

  GameObject? obj = getLetterObject(ch);
  if (obj == null) {
    await message("no such item.");
    return;
  }

  if (obj.whatIs != Cell.food) {
    await message("you can't eat that");
    return;
  }

  int moves = getRand(800, 1000);
  if (moves >= 900) {
    await message("yum, that tasted good");
  } else {
    await message("yuk, that food tasted awful");
    await addExp(3);
  }

  rogue.movesLeft ~/= 2;
  rogue.movesLeft += moves;
  hungerStr = "";
  printStats();

  await vanish(obj, true);
}

Future<void> holdMonster() async {
  int mcount = 0;

  for (int i = -2; i < 3; i++) {
    for (int j = -2; j < 3; j++) {
      int row = rogue.row + i;
      int col = rogue.col + j;

      if (row < minRow || row > ui.rows - 2 || col < 0 || col > ui.cols - 1) {
        continue;
      }

      if (screen[row][col] & Cell.monster != 0) {
        GameObject monster = objectAt(levelMonsters, row, col)!;
        monster.mFlags |= MonsterFlags.isAsleep;
        monster.mFlags &= ~MonsterFlags.wakens;
        mcount += 1;
      }
    }
  }

  if (mcount == 0) {
    await message("you feel a strange sense of loss");
  } else if (mcount == 1) {
    await message("the monster freezes");
  } else {
    await message("the monsters around you freeze");
  }
}

void teleport() {
  if (currentRoom >= 0) {
    darkenRoom(currentRoom);
  } else {
    ui.move(rogue.row, rogue.col);
    ui.write(getRoomChar(screen[rogue.row][rogue.col], rogue.row, rogue.col));
  }

  putPlayer();
  lightUpRoom();
  beingHeld = false;
}

void hallucinate() {
  if (blind != 0) {
    return;
  }

  for (GameObject obj in levelObjects) {
    String ch = ui.read(obj.row, obj.col, 1);

    if (!ch.between('A', 'Z') &&
        (obj.row != rogue.row || obj.col != rogue.col)) {
      if (ch != ' ' && ch != '.' && ch != '#' && ch != '+') {
        ui.move(obj.row, obj.col);
        ui.write(getRandObjChar());
      }
    }
  }

  for (GameObject obj in levelMonsters) {
    String ch = ui.read(obj.row, obj.col, 1);

    if (ch.between('A', 'Z')) {
      ui.move(obj.row, obj.col);
      ui.write(String.fromCharCode(getRand('A'.ascii, 'Z'.ascii)));
    }
  }
}

Future<void> unhallucinate() async {
  halluc = 0;

  if (currentRoom == passage) {
    lightPassage(rogue.row, rogue.col);
  } else {
    lightUpRoom();
  }

  await message("everything looks SO boring now");
}

Future<void> unblind() async {
  blind = 0;
  await message("the veil of darkness lifts");

  if (currentRoom == passage) {
    lightPassage(rogue.row, rogue.col);
  } else {
    lightUpRoom();
  }

  if (detectMonster) {
    showMonsters();
  }

  if (halluc != 0) {
    hallucinate();
  }
}

Future<void> sleepScroll() async {
  await message("you fall asleep");

  int i = getRand(4, 10);
  while (i > 0) {
    await moveMonsters();
    i -= 1;
  }

  await message("you can move again");
}

Future<void> goBlind() async {
  if (blind == 0) {
    await message("a cloak of darkness falls around you");
  }

  blind += getRand(500, 800);

  if (currentRoom >= 0) {
    Room r = rooms[currentRoom];

    for (int i = r.topRow + 1; i < r.bottomRow; i++) {
      for (int j = r.leftCol + 1; j < r.rightCol; j++) {
        ui.move(i, j);
        ui.write(' ');
      }
    }
  }

  ui.move(rogue.row, rogue.col);
  ui.write(rogue.fchar);
  ui.refresh();
}

String getEnchColor() {
  if (halluc != 0) {
    return idPotions[getRand(0, PotionType.values.length - 1)].title;
  }
  return "blue ";
}

void confuse() {
  confused = getRand(12, 22);
}

Future<void> unconfuse() async {
  confused = 0;
  await message("you feel less ${halluc != 0 ? 'trippy' : 'confused'} now");
}
